File: src-kotlin/task-manager-infrastructure/src/main/kotlin/com/example/taskmanager/infrastructure/InMemoryTaskRepository.kt
package com.example.taskmanager.infrastructure

import com.example.taskmanager.domain.tasks.Task
import com.example.taskmanager.domain.tasks.TaskId
import com.example.taskmanager.domain.tasks.TaskRepository
import com.example.taskmanager.domain.tasks.TaskStatus
import java.util.LinkedHashMap

/**
 * Simple in-memory repository used in workshop labs and examples.
 */
class InMemoryTaskRepository : TaskRepository {
    private val tasks = LinkedHashMap<TaskId, Task>()

    override fun save(task: Task): Task {
        tasks[task.id] = task
        return task
    }

    override fun findById(id: TaskId): Task? = tasks[id]

    override fun findByStatus(status: TaskStatus): List<Task> =
        tasks.values.filter { it.status == status }

    override fun findAll(): List<Task> = tasks.values.toList()

    override fun deleteById(id: TaskId) {
        tasks.remove(id)
    }

    override fun existsById(id: TaskId): Boolean = tasks.containsKey(id)

    override fun count(): Long = tasks.size.toLong()

    override fun countByStatus(status: TaskStatus): Long =
        tasks.values.count { it.status == status }.toLong()
}

File: src-swift/Sources/TaskManagerInfrastructure/InMemoryTaskRepository.swift
import TaskManagerDomain

/// Simple in-memory repository used in workshop labs and examples.
public final class InMemoryTaskRepository: TaskRepository {
    private var tasksByID: [TaskID: Task] = [:]
    private var taskOrder: [TaskID] = []

    public init(seedTasks: [Task] = []) {
        seedTasks.forEach { _ = save($0) }
    }

    @discardableResult
    public func save(_ task: Task) -> Task {
        if tasksByID[task.id] == nil {
            taskOrder.append(task.id)
        }

        tasksByID[task.id] = task
        return task
    }

    public func findByID(_ id: TaskID) -> Task? {
        tasksByID[id]
    }

    public func findByStatus(_ status: TaskStatus) -> [Task] {
        findAll().filter { $0.status == status }
    }

    public func findAll() -> [Task] {
        taskOrder.compactMap { tasksByID[$0] }
    }

    public func deleteByID(_ id: TaskID) {
        tasksByID.removeValue(forKey: id)
        taskOrder.removeAll { $0 == id }
    }

    public func exists(_ id: TaskID) -> Bool {
        tasksByID[id] != nil
    }

    public func count() -> Int {
        tasksByID.count
    }

    public func count(status: TaskStatus) -> Int {
        tasksByID.values.filter { $0.status == status }.count
    }
}
